File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/adapter/solution/AddressBookAdaptedToTable.java
// AddressBookAdaptedToTable.java

public class AddressBookAdaptedToTable extends Table
{
	public AddressBookAdaptedToTable(AddressBook aAddressBook)
	{
		if (aAddressBook == null)
		{
			throw new DpException("aAddressBook == null");
		}
		myAddressBook = aAddressBook;
	}

	public int getNumberOfRows()
	{
		return myAddressBook.getNumberOfPersons();
	}

	public int getNumberOfColumns()
	{
		return Person.getNumberOfFieldsPerPerson();
	}

	public String getColumnName(int aColumn)
	{
		return Person.getFieldName(aColumn);
	}

	public Object getValue(int aRow, int aColumn)
	{
		Person aPerson = myAddressBook.getPerson(aRow);
		return aPerson.getFieldValue(aColumn);
	}

	private AddressBook myAddressBook;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/adapter/lab/address_book_adapted_to_table.cpp
#include "address_book_adapted_to_table.h"

address_book_adapted_to_table::address_book_adapted_to_table()
{}

int address_book_adapted_to_table::get_number_of_rows()
{
	return my_address_book->get_number_of_persons();
}

int address_book_adapted_to_table::get_number_of_columns()
{
	return person::get_number_of_fields_per_person();
}